src/executor/evaluate/evaluated.rs

```diff
@@ -83,34 +83,46 @@ impl<'a> PartialOrd for Evaluated<'a> {
     }
 }
 
-macro_rules! binary_op {
-    ($name:ident, $op:tt) => {
-        pub fn $name<'b>(&self, other: &Evaluated<'a>) -> Result<Evaluated<'b>> {
-            let value_binary_op = |l: &Value, r: &Value| l.$name(r).map(Evaluated::from);
-
-            match (self, other) {
-                (Evaluated::Literal(l), Evaluated::Literal(r)) => {
-                    l.$name(r).map(Evaluated::Literal)
-                }
-                (Evaluated::Literal(l), Evaluated::Value(r)) => {
-                    value_binary_op(&Value::try_from(l)?, r.as_ref())
-                }
-                (Evaluated::Value(l), Evaluated::Literal(r)) => {
-                    value_binary_op(l.as_ref(), &Value::try_from(r)?)
-                }
-                (Evaluated::Value(l), Evaluated::Value(r)) => {
-                    value_binary_op(l.as_ref(), r.as_ref())
-                }
-            }
+fn binary_op<'a, 'b, T, U>(
+    l: &Evaluated<'a>,
+    r: &Evaluated<'b>,
+    value_op: T,
+    literal_op: U,
+) -> Result<Evaluated<'b>>
+where
+    T: Fn(&Value, &Value) -> Result<Value>,
+    U: Fn(&Literal<'a>, &Literal<'b>) -> Result<Literal<'b>>,
```

Comment:
It looks like a function used only once.
I recommend [FnOnce](https://doc.rust-lang.org/stable/std/ops/trait.FnOnce.html).

Reply:
Oh, I missed that, thanks!